telegram/ext/_picklepersistence.py

```diff
@@ -17,23 +17,119 @@
+    converted back into a class. Functions the same as :meth:`TelegramObject.__setstate__`.
+    This function should be kept in place for backwards compatibility even if the pickling logic
+    is changed, since `_custom_reduction` places references to this function into the pickled data.
+    """
+    bot = kwargs.pop('_bot', None)
+    obj = cls.__new__(cls)
+    for key, val in kwargs.items():
+        setattr(obj, key, val)
+    obj.set_bot(bot)
+    return obj  # type: ignore[return-value]
+
+
+def _custom_reduction(cls: TO) -> Tuple[Callable, Tuple[Type[TO], dict]]:
+    """
+    This method is used for pickling. The bot attribute is preserved so _BotPickler().persistent_id
+    works as intended.
+    """
+    data = cls._get_attrs(include_private=True)  # pylint: disable=protected-access
+    return _reconstruct_to, (cls.__class__, data)
+
+
+class _BotPickler(pickle.Pickler):
+    __slots__ = ('_bot',)
+
+    def __init__(self, bot: Bot, *args: Any, **kwargs: Any):
+        self._bot = bot
+        # Here we define a private dispatch_table, because we want to preserve the bot attribute of
+        # objects so persistent_id works as intended. Otherwise, the bot attribute is deleted in
+        # __getstate__, which is used during regular pickling (via pickle.dumps(...))
+        if py_ver < (3, 8):  # self.reducer_override is used above this version
+            self.dispatch_table = copyreg.dispatch_table.copy()  # type: ignore[attr-defined]
+            for obj in _all_subclasses(TelegramObject):
+                self.dispatch_table[obj] = _custom_reduction  # type: ignore[index]
+        super().__init__(*args, **kwargs)
+
+    def reducer_override(  # pylint: disable=no-self-use
+        self, obj: TO
+    ) -> Tuple[Callable, Tuple[Type[TO], dict]]:
+        if not isinstance(obj, TelegramObject):
+            return NotImplemented
+
+        return _custom_reduction(obj)
+
+    def persistent_id(self, obj: object) -> Optional[str]:
+        """Used to 'mark' the Bot, so it can be replaced later. See
+        https://docs.python.org/3/library/pickle.html#pickle.Pickler.persistent_id for more info
+        """
+        if obj is self._bot:
+            return _REPLACED_KNOWN_BOT
+        if isinstance(obj, Bot):
+            warn(
+                'Unknown bot instance found. Will be replaced by `None` during unpickling',
+                stacklevel=7,
+            )
+            return _REPLACED_UNKNOWN_BOT
+        return None  # pickles as usual
+
+
+class _BotUnpickler(pickle.Unpickler):
+    __slots__ = ('_bot',)
+
+    def __init__(self, bot: Bot, *args: Any, **kwargs: Any):
+        self._bot = bot
+        super().__init__(*args, **kwargs)
+
+    def persistent_load(self, pid: str) -> Optional[Bot]:
+        """Replaces the bot with the current bot if known, else it is replaced by :obj:`None`."""
+        return self._bot if pid == _REPLACED_KNOWN_BOT else None
```

Comment:
Can you raise `UnpicklingError` if the pid matches neither of `_R_(UN)KNOWN_BOT`? just to be sure …